[SPARK-17931] Eliminate unncessary task (de) serialization

kayousterhout opened the pull request:
## What changes were proposed in this pull request?

In the existing code, there are three layers of serialization
    involved in sending a task from the scheduler to an executor:
        - A Task object is serialized
        - The Task object is copied to a byte buffer that also
          contains serialized information about any additional JARs,
          files, and Properties needed for the task to execute. This
          byte buffer is stored as the member variable serializedTask
          in the TaskDescription class.
        - The TaskDescription is serialized (in addition to the serialized
          task + JARs, the TaskDescription class contains the task ID and
          other metadata) and sent in a LaunchTask message.
    
While it *is* necessary to have two layers of serialization, so that
the JAR, file, and Property info can be deserialized prior to
deserializing the Task object, the third layer of deserialization is
unnecessary.  This commit eliminates a layer of serialization by moving
the JARs, files, and Properties into the TaskDescription class.

This commit also serializes the Properties manually (by traversing the map),
as is done with the JARs and files, which reduces the final serialized size.

## How was this patch tested?

Unit tests

## Performance benefit and comparison with #15505 

This is a simpler alternative to the approach proposed in #15505.

@shivaram and I did some benchmarking of this and #15505 on a 20-machine m2.4xlarge EC2 machines (160 cores). We ran ~30 trials of code [1] (a very simple job with 10K tasks per stage) and measured the average time per stage:

Before this change: 2490ms
With this change: 2345 ms (so ~6% improvement over the baseline)
With @witgo's approach in #15505: 2046 ms (~18% improvement over baseline)

The reason that #15505 has a more significant improvement is that it also moves the serialization from the TaskSchedulerImpl thread to the CoarseGrainedSchedulerBackend thread. I added that functionality on top of this change, and got almost the same improvement [1] as #15505 (average of 2103ms). I think we should decouple these two changes, both so we have some record of the improvement form each individual improvement, and because this change is more about simplifying the code base (the improvement is negligible) while the other is about performance improvement.  The plan, currently, is to merge this PR and then merge the remaining part of #15505 that moves serialization.

[1] The reason the improvement wasn't quite as good as with #15505 when we ran the benchmarks is almost certainly because, at the point when we ran the benchmarks, I hadn't updated the code to manually serialize the Properties (instead the code was using Java's default serialization for the Properties object, whereas #15505 manually serialized the Properties).  This PR has since been updated to manually serialize the Properties, just like the other maps.

kayousterhout committed:
[SPARK-17931] Eliminate unncessary task (de) serialization

In the existing code, there are three layers of serialization
involved in sending a task from the scheduler to an executor:
    - A Task object is serialized
    - The Task object is copied to a byte buffer that also
      contains serialized information about any additional JARs,
      files, and Properties needed for the task to execute. This
      byte buffer is stored as the member variable serializedTask
      in the TaskDescription class.
    - The TaskDescription is serialized (in addition to the serialized
      task + JARs, the TaskDescription class contains the task ID and
      other metadata) and sent in a LaunchTask message.

While it *is* necessary to have two layers of serialization, so that
the JAR, file, and Property info can be deserialized prior to
deserializing the Task object, the third layer of deserialization is
unnecessary.  This commit eliminates a layer of serialization by moving
the JARs, files, and Properties into the TaskDescription class.

SparkQA commented:
**[Test build #69301 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69301/consoleFull)** for PR 16053 at commit [`e3cb71e`](https://github.com/apache/spark/commit/e3cb71ea57e135975ce3e28caaba314a9a599b86).

SparkQA commented:
**[Test build #69301 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69301/consoleFull)** for PR 16053 at commit [`e3cb71e`](https://github.com/apache/spark/commit/e3cb71ea57e135975ce3e28caaba314a9a599b86).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69301/
Test FAILed.

SparkQA commented:
**[Test build #69601 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69601/consoleFull)** for PR 16053 at commit [`8e2de02`](https://github.com/apache/spark/commit/8e2de02f7e365403e2c551fa035df766a8f12d39).

SparkQA commented:
**[Test build #69603 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69603/consoleFull)** for PR 16053 at commit [`063ff52`](https://github.com/apache/spark/commit/063ff52140962c0fcae3a2806959318aa3684d6c).

SparkQA commented:
**[Test build #69601 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69601/consoleFull)** for PR 16053 at commit [`8e2de02`](https://github.com/apache/spark/commit/8e2de02f7e365403e2c551fa035df766a8f12d39).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `  case class LaunchTask(`

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69601/
Test FAILed.

SparkQA commented:
**[Test build #69603 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69603/consoleFull)** for PR 16053 at commit [`063ff52`](https://github.com/apache/spark/commit/063ff52140962c0fcae3a2806959318aa3684d6c).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69603/
Test FAILed.

SparkQA commented:
**[Test build #69640 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69640/consoleFull)** for PR 16053 at commit [`1917e35`](https://github.com/apache/spark/commit/1917e3546df9cb6967d6df1b7e975e9c86ca3bc7).

SparkQA commented:
**[Test build #69640 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69640/consoleFull)** for PR 16053 at commit [`1917e35`](https://github.com/apache/spark/commit/1917e3546df9cb6967d6df1b7e975e9c86ca3bc7).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69640/
Test FAILed.

SparkQA commented:
**[Test build #69643 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69643/consoleFull)** for PR 16053 at commit [`e1ad96b`](https://github.com/apache/spark/commit/e1ad96bdf3c4049ef9943f2f6d1534114da3ec2c).

SparkQA commented:
**[Test build #69644 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69644/consoleFull)** for PR 16053 at commit [`13bacfe`](https://github.com/apache/spark/commit/13bacfeecd84376f102d706612e3a244d17330f3).

SparkQA commented:
**[Test build #69643 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69643/consoleFull)** for PR 16053 at commit [`e1ad96b`](https://github.com/apache/spark/commit/e1ad96bdf3c4049ef9943f2f6d1534114da3ec2c).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69643/
Test FAILed.

SparkQA commented:
**[Test build #69644 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69644/consoleFull)** for PR 16053 at commit [`13bacfe`](https://github.com/apache/spark/commit/13bacfeecd84376f102d706612e3a244d17330f3).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69644/
Test FAILed.

SparkQA commented:
**[Test build #69646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69646/consoleFull)** for PR 16053 at commit [`091e0d0`](https://github.com/apache/spark/commit/091e0d01adcf895e5ffefb631291290217302211).

SparkQA commented:
**[Test build #69646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69646/consoleFull)** for PR 16053 at commit [`091e0d0`](https://github.com/apache/spark/commit/091e0d01adcf895e5ffefb631291290217302211).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69646/
Test FAILed.

SparkQA commented:
**[Test build #69685 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69685/consoleFull)** for PR 16053 at commit [`ab0bcf2`](https://github.com/apache/spark/commit/ab0bcf20c6b04d334b335551e689164c5a21cd04).

SparkQA commented:
**[Test build #69685 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69685/consoleFull)** for PR 16053 at commit [`ab0bcf2`](https://github.com/apache/spark/commit/ab0bcf20c6b04d334b335551e689164c5a21cd04).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69685/
Test FAILed.

SparkQA commented:
**[Test build #69701 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69701/consoleFull)** for PR 16053 at commit [`0d3d656`](https://github.com/apache/spark/commit/0d3d6561a0556923f804eaff6f661a5082fe2fcc).

SparkQA commented:
**[Test build #69701 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69701/consoleFull)** for PR 16053 at commit [`0d3d656`](https://github.com/apache/spark/commit/0d3d6561a0556923f804eaff6f661a5082fe2fcc).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69701/
Test FAILed.

SparkQA commented:
**[Test build #69702 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69702/consoleFull)** for PR 16053 at commit [`e3cb71e`](https://github.com/apache/spark/commit/e3cb71ea57e135975ce3e28caaba314a9a599b86).

SparkQA commented:
**[Test build #69702 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69702/consoleFull)** for PR 16053 at commit [`e3cb71e`](https://github.com/apache/spark/commit/e3cb71ea57e135975ce3e28caaba314a9a599b86).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69702/
Test FAILed.

kayousterhout committed:
Manually serialize task description.

kayousterhout committed:
Added a test for TaskDescription serialization

kayousterhout committed:
Partial implementation of Mesos functionality.

This commit will break behavior with Mesos -- submitting this
because it doesn't appear there are any end-to-end tests for
launching Tasks with Mesos, and I want to double check whether
that's true.

kayousterhout committed:
Merge remote-tracking branch 'upstream/master' into SPARK-17931

SparkQA commented:
**[Test build #70277 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70277/testReport)** for PR 16053 at commit [`6a30d52`](https://github.com/apache/spark/commit/6a30d52f36acaa6bfc429b14aea5118d820272c1).

SparkQA commented:
**[Test build #70277 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70277/testReport)** for PR 16053 at commit [`6a30d52`](https://github.com/apache/spark/commit/6a30d52f36acaa6bfc429b14aea5118d820272c1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70277/
Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/scheduler/TaskDescription.scala:
```diff
@@ -17,27 +17,119 @@
+ *         important. For this reason, we serialize TaskDescriptions ourselves with the
+ *         TaskDescription.encode and TaskDescription.decode methods.  This results in a smaller
+ *         serialized size because it avoids serializing unnecessary fields in the Map objects
+ *         (which can introduce significant overhead when the maps are small).
```
I realize you are just adding comments to explain the old behavior -- if this explanation about serializing maps is right, then shouldn't you do the same for the `properties`?

kayousterhout replied:
Oh this is a great point, and I realize this is likely also the explanation for the ~2% remaining performance disparity between this change and #15505 that I mentioned in my comment.  Thanks for noticing this oversight -- I'll fix this!

squito reviewed (commented):
just one minor comment -- I will leave my more general comments about comparison of approaches on https://github.com/apache/spark/pull/15505

kayousterhout commented:
Thanks for the review @squito.  I got sidetracked from this at the end of last week and forgot to post the results of some benchmarks @shivaram and I did on a 20-machine m2.4xlarge EC2 machines (160 cores).  We ran ~30 trials of code [1] (a very simple job with 10K tasks per stage) and measured the average time per stage:

Before this change: 2490ms
With this change: 2345 ms (so ~6% improvement over the baseline)
With @witgo's approach in #15505: 2046 ms (~18% improvement over baseline)

The reason that #15505 has a more significant improvement is that it also moves the serialization from the TaskSchedulerImpl thread to the CoarseGrainedSchedulerBackend thread.  I added that functionality on top of this change, and got almost the same improvement as #15505 (average of 2103ms).  I think we should decouple these two changes, both so we have some record of the improvement form each individual improvement, and because this change is more about simplifying the code base (the improvement is negligible) while the other is about performance improvement.  I filed a separate JIRA for that issue here: https://issues.apache.org/jira/browse/SPARK-18890

kayousterhout commented:
Also I need to fix up the Mesos functionality here.  I deliberately posted a broken Mesos version to verify my hypothesis that there are no end-to-end Mesos tests (this code would have failed any such tests), and need to fix it up and then ask one of the Mesos folks to take a look to double check things, since the last of end-to-end tests is a bit scary.

kayousterhout committed:
Finished Mesos clean-up.

Removed MesosTaskLaunchData, since now we can just use the
serialized task description as the means of passing data to
Mesos (this eliminates code and is also more consistent with
how tasks are launched in Mesos coarse-grained mode and with
Spark's standalone scheudler).

kayousterhout committed:
Serialize properties manually.

Since Properties is a map, like the jars and files, it's
more efficient to serialize it manually to avoid needing
to serialize extra member variables of the Properties
object (for the simple example in TaskDescriptionSuite,
this change reduces the size of the serialized
TaskDescription from 309 bytes to 149 bytes).

SparkQA commented:
**[Test build #70373 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70373/testReport)** for PR 16053 at commit [`ecd0731`](https://github.com/apache/spark/commit/ecd07311d21221e0d862d39d582922f0eab359d4).

SparkQA commented:
**[Test build #70373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70373/testReport)** for PR 16053 at commit [`ecd0731`](https://github.com/apache/spark/commit/ecd07311d21221e0d862d39d582922f0eab359d4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70373/
Test PASSed.

kayousterhout commented:
I fixed all of the issues mentioned with the commits I pushed a few days ago, in case anyone was waiting for that and didn't notice!

shivaram commented:
@kayousterhout it shows there are some conflicts ?

shivaram commented:
I think the code change in core/ looks pretty good to me. Is there someone who can look at the mesos changes ?

mridulm commented on core/src/main/scala/org/apache/spark/scheduler/TaskDescription.scala:
```diff
@@ -17,27 +17,123 @@
 /**
  * Description of a task that gets passed onto executors to be executed, usually created by
- * `TaskSetManager.resourceOffer`.
+  * `TaskSetManager.resourceOffer`.
```
nit: indent

kayousterhout replied:
Done

mridulm commented on core/src/main/scala/org/apache/spark/scheduler/TaskDescription.scala:
```diff
@@ -17,27 +17,123 @@
+    for ((name, timestamp) <- taskDescription.addedJars) {
+      dataOut.writeUTF(name)
+      dataOut.writeLong(timestamp)
+    }
```
nit: Looks like repeated code which could be refactored out as a single method ?

```
def writeStringLongMap(map: Map[String, Long]): Unit = {
  dataOut.writeInt(map.size)
  for ((k, v) <- map) {
    dataOut.writeUTF(k)
    dataOut.writeLong(v)
  }
}

writeStringLongMap(taskDescription.addedFiles)
writeStringLongMap(taskDescription. addedJars)

```

Similar applies to decode.
The reason I mention is that both properties have very similar names, and it is easy to misread the code as it evolves.

mridulm commented on core/src/main/scala/org/apache/spark/scheduler/TaskDescription.scala:
```diff
@@ -17,27 +17,123 @@
+    taskDescription.properties.stringPropertyNames.asScala.foreach { name =>
+      dataOut.writeUTF(name)
+      dataOut.writeUTF(taskDescription.properties.getProperty(name))
+    }
```
Why not :
```
taskDescription.properties.asScala.foreach { case (key, value) => {
    dataOut.writeUTF(key)
    dataOut.writeUTF(value)
  }
}
```

Avoid probing the map again.

mridulm commented on core/src/main/scala/org/apache/spark/scheduler/TaskDescription.scala:
```diff
@@ -17,27 +17,123 @@
+    // (this requires first flushing the data output stream, so that all of the data has been
+    // written from the data output stream so the underlying ByteBufferOutputStream before
+    // we write the task).
+    dataOut.flush()
```
This is a noop and can be removed.

mridulm commented on core/src/main/scala/org/apache/spark/scheduler/TaskDescription.scala:
```diff
@@ -17,27 +17,123 @@
+    // written from the data output stream so the underlying ByteBufferOutputStream before
+    // we write the task).
+    dataOut.flush()
+    Utils.writeByteBuffer(taskDescription.serializedTask, bytesOut)
```
Just noticed that `Utils.writeByteBuffer` actually can have different state for serializedTask on return.
If it has array, `remaining` remains unchanged - if not, `remaining` will be 0 : was this on purpose ?

Will this behavior have issues in our codepath here ? (needs to be fixed in Utils.writeByteBuffer - but unrelated to this pr ?)
Or perhaps send `taskDescription.serializedTask.duplicate()` over to the method ?

kayousterhout replied:
Just opened #16462 to fix this.  I don't think this has any issues for this code path, because this is exactly the same as the old code that was previously located in Task.scala (which was moved here in this PR).

mridulm reviewed (commented):
Added a few comments.

kayousterhout committed:
Merge remote-tracking branch 'apache/master' into SPARK-17931

kayousterhout committed:
Addressed review comments from @mridulm

SparkQA commented:
**[Test build #70827 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70827/testReport)** for PR 16053 at commit [`8c17ebd`](https://github.com/apache/spark/commit/8c17ebd12334e360904e47cb5009be7c94c9b255).

kayousterhout committed:
Small cleanup

SparkQA commented:
**[Test build #70828 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70828/testReport)** for PR 16053 at commit [`db69955`](https://github.com/apache/spark/commit/db69955d9faeec4b408002cb054ed7fc57257d1e).

kayousterhout reviewed (commented):
@squito @mridulm @shivaram Thanks for looking at this! I've addressed all of the comments and opened a separate PR for one of the issues you mentioned Mridul.

@mgummelt would you mind taking a look at the Mesos changes here?  I've checked them over but there aren't any unit tests for this part of the code, so it would be great to have a second set of eyes (or, even better, if you happen to have a Mesos cluster up, would be awesome to quickly test that tasks can still be launched properly in coarse grained mode).

SparkQA commented:
**[Test build #70828 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70828/testReport)** for PR 16053 at commit [`db69955`](https://github.com/apache/spark/commit/db69955d9faeec4b408002cb054ed7fc57257d1e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70828/
Test FAILed.

kayousterhout commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #70835 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70835/testReport)** for PR 16053 at commit [`db69955`](https://github.com/apache/spark/commit/db69955d9faeec4b408002cb054ed7fc57257d1e).

SparkQA commented:
**[Test build #70827 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70827/testReport)** for PR 16053 at commit [`8c17ebd`](https://github.com/apache/spark/commit/8c17ebd12334e360904e47cb5009be7c94c9b255).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70827/
Test PASSed.

SparkQA commented:
**[Test build #70835 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70835/testReport)** for PR 16053 at commit [`db69955`](https://github.com/apache/spark/commit/db69955d9faeec4b408002cb054ed7fc57257d1e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70835/
Test FAILed.

kayousterhout committed:
Cleanup visibility of TaskRunner member vars

kayousterhout committed:
Tiny cleanup

SparkQA commented:
**[Test build #70842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70842/testReport)** for PR 16053 at commit [`c983c8f`](https://github.com/apache/spark/commit/c983c8f0eafc4a9f7823d347feef572ec30e2b4a).

SparkQA commented:
**[Test build #70842 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70842/testReport)** for PR 16053 at commit [`c983c8f`](https://github.com/apache/spark/commit/c983c8f0eafc4a9f7823d347feef572ec30e2b4a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70842/
Test FAILed.

mgummelt commented:
@kayousterhout It looks like you only touched deprecated Mesos code (fine-grained mode), so LGTM

kayousterhout commented:
Cool thanks for looking at the Mesos changes here @mgummelt!

squito commented:
Jenkins, retest this please

squito commented:
lgtm, assuming tests pass

kayousterhout commented:
Cool thanks for the review @squito and for re-testing (I tracked down the flaky test that caused this but then forgot to re-test!)

SparkQA commented:
**[Test build #70941 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70941/testReport)** for PR 16053 at commit [`c983c8f`](https://github.com/apache/spark/commit/c983c8f0eafc4a9f7823d347feef572ec30e2b4a).

SparkQA commented:
**[Test build #70941 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70941/testReport)** for PR 16053 at commit [`c983c8f`](https://github.com/apache/spark/commit/c983c8f0eafc4a9f7823d347feef572ec30e2b4a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70941/
Test PASSed.

squito commented:
merged to master
